File: tests/snapshot_pre_upgrade/space_station.go
package main

import (
	"fmt"
	"roundtrip/builtins_py"
)

func Contains[T comparable](slice []T, item T) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

func Range(start, end int) []int {
	var r []int
	for i := start; i < end; i++ {
		r = append(r, i)
	}
	return r
}

func Ptr[T any](v T) *T {
	return &v
}

type SpaceStation struct {

	name builtins_py.String

	capacity int

	modules builtins_py.List_str

	resource_levels builtins_py.Map_str_int

}

func NewSpaceStation(name builtins_py.String, capacity int) *SpaceStation {
	this := &SpaceStation{}
	this.name = name
	this.capacity = capacity
	this.modules = builtins_py.List_str{builtins_py.String("Command"), builtins_py.String("Life Support"), builtins_py.String("Laboratory")}
	this.resource_levels = builtins_py.Map_str_int{builtins_py.String("Oxygen"): 95, builtins_py.String("Power"): 100}
	return this
}

func (this *SpaceStation) Printstatus() {
	fmt.Println(builtins_py.String("Station: ") + this.name)
	fmt.Println(builtins_py.String(fmt.Sprintf("%v", builtins_py.String("Capacity: "))) + builtins_py.String(fmt.Sprintf("%v", this.capacity)))
}
type Astronaut struct {

	name builtins_py.String

	experience int

	assigned_station Optional[SpaceStation]

	health int

}

func NewAstronaut(name builtins_py.String, experience int, assigned_station Optional[SpaceStation]) *Astronaut {
	this := &Astronaut{}
	this.name = name
	this.experience = experience
	this.assigned_station = assigned_station
	this.health = 100
	return this
}

func (this *Astronaut) Work(hours int) int {
		if hours < 0 {
		panic(NewException(builtins_py.String("Cannot work negative hours")))
	}
	var tasks_done int = 0
	var current_hour int = 0
		for current_hour < hours {
				if this.health < 20 {
			fmt.Println(builtins_py.String("Astronaut is too tired to work."))
		}
				if current_hour % 3 == 0 {
			current_hour = current_hour + 1
		}
		this.health = this.health - 10
		tasks_done = tasks_done + 2
		current_hour = current_hour + 1
	}
	return tasks_done
	return 0
}

func (this *Astronaut) Report() {
		if this.assigned_station != nil {
		this.assigned_station.Printstatus()
				if len(this.assigned_station.modules) > 0 {
			fmt.Println(builtins_py.String(fmt.Sprintf("%v", builtins_py.String("Primary module: "))) + builtins_py.String(fmt.Sprintf("%v", this.assigned_station.modules[0])))
		}
				if Contains(this.assigned_station.resource_levels, builtins_py.String("Oxygen")) {
			var oxy int = this.assigned_station.resource_levels[builtins_py.String("Oxygen")]
			fmt.Println(builtins_py.String(fmt.Sprintf("%v", builtins_py.String("Oxygen Level: "))) + builtins_py.String(fmt.Sprintf("%v", oxy)))
		}
	} else {
		fmt.Println(builtins_py.String("No station assigned."))
	}
}
func main() {
		func() {
		defer func() {
			if e := recover(); e != nil {
			}
		}()
		var station *SpaceStation = NewSpaceStation(builtins_py.String("Alpha"), 50)
		var astro *Astronaut = NewAstronaut(builtins_py.String("Ripley"), 10, station)
		astro.Report()
		var completed int = astro.Work(8)
		fmt.Println(builtins_py.String(fmt.Sprintf("%v", builtins_py.String("Tasks Completed: "))) + builtins_py.String(fmt.Sprintf("%v", completed)))
		fmt.Println(builtins_py.String(fmt.Sprintf("%v", builtins_py.String("Health remaining: "))) + builtins_py.String(fmt.Sprintf("%v", astro.health)))
		var rookie *Astronaut = NewAstronaut(builtins_py.String("Newb"), 0, nil)
		rookie.Report()
		rookie.Work( - 5)
	}()
}

File: tests/snapshot_pre_upgrade/space_station.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <any>
#include <stdexcept>

class SpaceStation {
public:
    std::string name;
    int capacity;
    std::vector<std::string> modules;
    std::unordered_map<std::string, int> resource_levels;
    SpaceStation(std::string name, int capacity) {
        this->name = name;
        this->capacity = capacity;
        this->modules = {"Command", "Life Support", "Laboratory"};
        this->resource_levels = builtins_py::hashmap({{"Oxygen", 95}, {"Power", 100}});
    }
    void printStatus() {
        std::cout << ("Station: " + this->name) << "\n";
        std::cout << (builtins_py::str("Capacity: ") + builtins_py::str(this->capacity)) << "\n";
    }
};
class Astronaut {
public:
    std::string name;
    int experience;
    Optional[SpaceStation] assigned_station;
    int health;
    Astronaut(std::string name, int experience, Optional[SpaceStation] assigned_station) {
        this->name = name;
        this->experience = experience;
        this->assigned_station = assigned_station;
        this->health = 100;
    }
    int work(int hours) {
        if ((hours < 0)) {
            throw std::runtime_error(Exception("Cannot work negative hours"));
        }
        tasks_done = 0;
        current_hour = 0;
        while ((current_hour < hours)) {
            if ((this->health < 20)) {
                std::cout << "Astronaut is too tired to work." << "\n";
            }
            if (((current_hour % 3) == 0)) {
                current_hour = (current_hour + 1);
            }
            this->health = (this->health - 10);
            tasks_done = (tasks_done + 2);
            current_hour = (current_hour + 1);
        }
        return tasks_done;
    }
    void report() {
        if ((this->assigned_station != null)) {
            this->assigned_station.printStatus();
            if ((builtins_py::len(this->assigned_station.modules) > 0)) {
                std::cout << (builtins_py::str("Primary module: ") + builtins_py::str(this->assigned_station.modules[0])) << "\n";
            }
            if (("Oxygen" in this->assigned_station.resource_levels)) {
                oxy = this->assigned_station.resource_levels["Oxygen"];
                std::cout << (builtins_py::str("Oxygen Level: ") + builtins_py::str(oxy)) << "\n";
            }
        } else {
            std::cout << "No station assigned." << "\n";
        }
    }
};
int main() {
    try {
        SpaceStation station = SpaceStation("Alpha", 50);
        Astronaut astro = Astronaut("Ripley", 10, station);
        astro.report();
        int completed = astro.work(8);
        std::cout << (builtins_py::str("Tasks Completed: ") + builtins_py::str(completed)) << "\n";
        std::cout << (builtins_py::str("Health remaining: ") + builtins_py::str(astro.health)) << "\n";
        Astronaut rookie = Astronaut("Newb", 0, null);
        rookie.report();
        rookie.work(-(5));
    } catch (const std::exception& e) {
    }
    return 0;
}
if ((__name__ == "__main__")) {
    main();
}